### PR TITLE
Blog App Milestone #8: Integration specs for Views and fixing n+1 problems.

### PR DESCRIPTION
**Pull request by karlavdelgadof**
## Blog App Milestone #8: Integration specs for Views and fixing n+1 problems :hammer_and_wrench: :memo:  :heavy_check_mark: 

These are the proposed changes :recycle:  for this project :arrow_down: :
### n+1
- Implemented eager loading to fix the N+1 problem when fetching all posts and their comments for a user :100: 

### Integration specs :1234: 

- Used Capybara to write integration tests for each view :+1: : 
    - Created tests for User index page :busts_in_silhouette: 
    - Created tests for User show page :bust_in_silhouette: 
    - Created tests for the User posts index page :spiral_notepad: :bust_in_silhouette: 
    - Created tests for Post show page :spiral_notepad: 

SCREENSHOT WITH REQUIRED TEST
image

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 class PostsController < ApplicationController
   def index
-    @posts = Post.all
-    @user = User.includes(posts: [:comments]).find(params[:user_id].to_i)
+    @user = User.find(params[:user_id].to_i)
+    @posts = @user.posts.includes([:author])
   end
```
Comment (omar-muhamad) [inline app/controllers/posts_controller.rb:5]: - [x] To fix `n+1` problem, Please make line 4 like this:
```ruby
@user.posts.includes(:comments)
```

File: spec/features/user_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    end
+    it 'should see the number of posts each user has written' do
+      @users.each do |user|
+        expect(page).to have_content(user.post_counter)
+      end
+    end
+    it 'should redirect to users show page' do
+      visit user_path(@users.ids)
+      expect(page).to have_content(@user.name)
+      expect(page).to have_content(@user.bio)
```
Comment (omar-muhamad) [inline spec/features/user_index_spec.rb]: - [x] To make the test inspect the redirection to the user show page please use this matcher `have_current_path` like this:
```ruby
expect(page).to have_current_path(user_path(@user.id))
```

File: spec/features/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    it 'should show the first comments on a post' do
+      expect(page).to have_content(@post.comments.first)
+    end
+    it 'should show how many comments a post has.' do
+      expect(page).to have_content(@post.comments_counter)
+    end
+    it 'should show how many likes a post has.' do
+      expect(page).to have_content(@post.likes_counter)
+    end
+  end
```
Comment (omar-muhamad) [inline spec/features/post_index_spec.rb]: - [x] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=I%20can%20see%20a%20section%20for%20pagination%20if%20there%20are%20more%20posts%20than%20fit%20on%20the%20view.), Please add a test that asserts you can see a section for pagination if there are more posts than fit on the view.
- [x] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20on%20a%20post%2C%20it%20redirects%20me%20to%20that%20post%27s%20show%20page.). Please add a Test that redirects to the post's show page when you click on a post.

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @karlavdelgadof and @NATASHA-ct,

Nice job, you are almost there.



## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used the correct workflow.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [x] Please check the inline comments below.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (NATASHA-ct) [reply app/controllers/posts_controller.rb:5]: Thank you for your review but this suggestion is creating a bullet error for me.
Currently I don't have any N+1 error  

image

**Review by SajjadAhmad14 — APPROVED** [review]
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:

### Hi Tascha, 
## To Highlight 🎉
Git flow is followed. :heavy_check_mark:
README is professional. :heavy_check_mark:
## Congratulations 🎉 Thank you for implementing all the requirements 🙏
### Please go ahead and merge the branch!

## Happy Coding 👏👏👏
#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
#### Please remember to tag me in your question so I can receive the notification.
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**